File: leetcode/solved/53_Maximum_Subarray/solution.cpp
/*
 * solution.cpp
 * Copyright (C) 2017 Sungmin <smoh2044@gmail.com>
 *
 * Distributed under terms of the MIT license.
 */

#include <iostream>
#include <vector>

using namespace std;


class Solution {
    public:
        int maxSubArray(vector<int>& nums) {
            int until;
            int sofar;
            until = sofar = nums[0];
            int n = nums.size();
            for(int i=1; i<n; ++i){
                until = max(nums[i], until+nums[i]);
                sofar = max(until, sofar);
            }
            return sofar;
        }
};

int main(){
    ios_base::sync_with_stdio(false);
    int n;
    cin >> n;
    vector<int> nums(n);
    for(int i=0; i<n; ++i){
        cin >> nums[i];
    }
    int s = Solution().maxSubArray(nums);
    cout << s << endl;

    return 0;
}

File: leetcode/solved/53_Maximum_Subarray/solution.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2020 sungminoh <smoh2044@gmail.com>
#
# Distributed under terms of the MIT license.

"""
Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.

A subarray is a contiguous part of an array.

Example 1:

Input: nums = [-2,1,-3,4,-1,2,1,-5,4]
Output: 6
Explanation: [4,-1,2,1] has the largest sum = 6.

Example 2:

Input: nums = [1]
Output: 1

Example 3:

Input: nums = [5,4,-1,7,8]
Output: 23

Constraints:

	1 <= nums.length <= 105
	-104 <= nums[i] <= 104

Follow up: If you have figured out the O(n) solution, try coding another solution using the divide and conquer approach, which is more subtle.
"""
import sys
from typing import List
import pytest


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        """
        Time complexity: O(n)
        Space complexity: O(1)
        """
        ret = nums[0]
        prevsum = nums[0]
        for n in nums[1:]:
            prevsum = max(n, prevsum + n)
            ret = max(ret, prevsum)

        return ret


@pytest.mark.parametrize('nums, expected', [
    ([-2,1,-3,4,-1,2,1,-5,4], 6),
    ([1], 1),
    ([5,4,-1,7,8], 23),
])
def test(nums, expected):
    assert expected == Solution().maxSubArray(nums)


if __name__ == '__main__':
    sys.exit(pytest.main(["-s", "-v"] + sys.argv))
